app/server-frontend.js

```diff
@@ -1,15 +1,58 @@
 const express = require('express');
+const basicAuth = require('express-basic-auth');
+const otp = require('node-2fa');
 const path = require('path');
 const config = require('config');
+const fs = require('fs');
 
 const { configureWebSocket } = require('./websocket/configure');
 const { configureLocalTunnel } = require('./local-tunnel/configure');
 
+let OTP_SECRET = false;
+if (fs.existsSync('./.2fa_secret')) {
+  OTP_SECRET = fs.readFileSync('./.2fa_secret', { encoding: 'utf8' });
+}
+
+const basicAuthChallenge = (username, password) => {
+  if (!config.frontend_auth) {
+    return true;
+  }
+
+  const usernameOk = basicAuth.safeCompare(
+    username,
+    config.frontend_auth.username
+  );
+
+  let passwordOk = false;
+  if (config.frontend_auth.mode_otp && OTP_SECRET) {
+    const passwordPlain = basicAuth.safeCompare(password.slice(0, -6), config.frontend_auth.password);
+    const password2fa = Boolean(otp.verifyToken(OTP_SECRET, password.replace(config.frontend_auth.password, '')))
+
+    passwordOk = passwordPlain && password2fa;
+  } else {
+    passwordOk = basicAuth.safeCompare(password, config.frontend_auth.password);
+  }
+
+  return usernameOk & passwordOk;
+};
+
 const runFrontend = async serverLogger => {
   const logger = serverLogger.child({ server: 'frontend' });
   logger.info({ config }, `API ${config.get('mode')} frontend started on`);
 
   const app = express();
+  if (config.frontend_auth) {
+    const baOptions = {
+      authorizer: basicAuthChallenge,
+      challenge: true,
+      realm: 'Binance Trading Bot'
+    };
+    if (config.frontend_auth.mode_otp) {
```

Comment:
I did generate `.2fa_secret` and set the configuration as below:
![image](https://user-images.githubusercontent.com/5715919/111893673-6e774f00-8a58-11eb-9fca-576d892b13a1.png)

However, I still can access without adding OTP code to the password

![image](https://user-images.githubusercontent.com/5715919/111893677-7c2cd480-8a58-11eb-9714-66f3b46166f5.png)

![image](https://user-images.githubusercontent.com/5715919/111893683-818a1f00-8a58-11eb-95ca-99ffe23d6dea.png)

Reply:
Nothing.. it's a bug about the password/2fa length

Reply:
working now